File: julia8024/BOJ_11724.py
# 백준
# 11724. 연결 요소의 개수

import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline


def dfs(graph, v, visited):
    visited[v] = True
    for i in graph[v]:
        if not visited[i]:
            dfs(graph, i, visited)


# 정점의 개수 N, 간선의 개수 M
N, M = map(int, input().split())

graph = [[] for _ in range(N+1)]

for i in range(M):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

count = 0 # 연결 노드 수
visited = [False] * (N+1)
for i in range(1, N+1):
    if not visited[i]:
        dfs(graph, i, visited)
        count += 1 # dfs 한 번 끝날 때마다 count+1

print(count)

File: chaeryeon823/BOJ_11724.js
const fs = require("fs");
const input = fs.readFileSync(0, "utf8").trim().split("\n");

const [N, M] = input[0].split(" ").map(Number);

const graph = Array.from({ length: N + 1 }, () => []);
for (let i = 1; i <= M; i++) {
  const [u, v] = input[i].split(" ").map(Number);
  graph[u].push(v);
  graph[v].push(u);
}

const visited = Array(N + 1).fill(false);

function dfs(start) {
  const stack = [start];
  visited[start] = true;

  while (stack.length) {
    const node = stack.pop();
    for (const next of graph[node]) {
      if (!visited[next]) {
        visited[next] = true;
        stack.push(next);
      }
    }
  }
}

let count = 0;
for (let i = 1; i <= N; i++) {
  if (!visited[i]) {
    count++;
    dfs(i);
  }
}

console.log(count);
